### ISSUE TITLE
Error running pso-explorer

### ISSUE THREAD

**Issue by zapalotta** [PureStorage-OpenConnect/pso-explorer-container#22]
Hi,

I installed pso-explorer using hlem 2 on Openshift 3.11. Installation worked, pods are running but the UI keeps displaying "Please standby, while we are collecting data...", even the whole page refreshes after some seconds again and again.

The logs show:

`2021-03-17 13:24:39] local.INFO:        Call getStorageClasses()
[2021-03-17 13:24:39] local.INFO:        Call getPersistentVolumes()
[2021-03-17 13:24:39] local.ERROR: Invalid argument supplied for foreach() {"exception":"[object] (ErrorException(code: 0): Invalid argument supplied for foreach() at /var/www/app/Pso.php:1123)
[stacktrace]
#0 /var/www/app/Pso.php(1123): Illuminate\\Foundation\\Bootstrap\\HandleExceptions->handleError(2, 'Invalid argumen...', '/var/www/app/Ps...', 1123, Array)
#1 /var/www/app/Pso.php(1965): App\\Pso->getPersistentVolumes()
#2 /var/www/app/Pso.php(103): App\\Pso->refreshData()
#3 /var/www/app/Http/Controllers/NodeAgentController.php(26): App\\Pso->__construct()
#4 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): App\\Http\\Controllers\\NodeAgentController->getInfo(Object(App\\Http\\Requests\\NodeAgentGetRequest))
#5 /var/www/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\\Routing\\Controller->callAction('getInfo', Array)
#6 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Route.php(254): Illuminate\\Routing\\ControllerDispatcher->dispatch(Object(Illuminate\\Routing\\Route), Object(App\\Http\\Controllers\\NodeAgentController), 'getInfo')
#7 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Route.php(197): Illuminate\\Routing\\Route->runController()
#8 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(693): Illuminate\\Routing\\Route->run()
#9 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))
#10 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#11 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#12 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#13 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\VerifyCsrfToken->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#14 /var/www/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#15 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#16 /var/www/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#17 /var/www/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(63): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest(Object(Illuminate\\Http\\Request), Object(Illuminate\\Session\\Store), Object(Closure))
#18 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Session\\Middleware\\StartSession->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#19 /var/www/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#20 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#21 /var/www/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#22 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#23 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#24 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(695): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#25 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(670): Illuminate\\Routing\\Router->runRouteWithinStack(Object(Illuminate\\Routing\\Route), Object(Illuminate\\Http\\Request))
#26 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(636): Illuminate\\Routing\\Router->runRoute(Object(Illuminate\\Http\\Request), Object(Illuminate\\Routing\\Route))
#27 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(625): Illuminate\\Routing\\Router->dispatchToRoute(Object(Illuminate\\Http\\Request))
#28 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(166): Illuminate\\Routing\\Router->dispatch(Object(Illuminate\\Http\\Request))
#29 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}(Object(Illuminate\\Http\\Request))
#30 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#31 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#32 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#33 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#34 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#35 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#36 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#37 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#38 /var/www/vendor/fruitcake/laravel-cors/src/HandleCors.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#39 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\\Cors\\HandleCors->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#40 /var/www/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#41 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fideloper\\Proxy\\TrustProxies->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#42 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#43 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(141): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#44 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))
#45 /var/www/public/index.php(55): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#46 {main}
"}
127.0.0.1 -  17/Mar/2021:13:24:39 +0000 "GET /index.php" 500
[2021-03-17 13:24:39] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:39 +0000 "GET /index.php" 200
[2021-03-17 13:24:40] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:40 +0000 "GET /index.php" 200
[2021-03-17 13:24:40] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:40 +0000 "GET /index.php" 200
[2021-03-17 13:24:40] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:40 +0000 "GET /index.php" 200
[2021-03-17 13:24:43] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:43 +0000 "GET /index.php" 200
[2021-03-17 13:24:46] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:46 +0000 "GET /index.php" 200
[2021-03-17 13:24:49] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:49 +0000 "GET /index.php" 200
[2021-03-17 13:24:49] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:49 +0000 "GET /index.php" 200
[2021-03-17 13:24:50] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:50 +0000 "GET /index.php" 200
[2021-03-17 13:24:52] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:52 +0000 "GET /index.php" 200
[2021-03-17 13:24:56] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:56 +0000 "GET /index.php" 200
[2021-03-17 13:24:59] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:59 +0000 "GET /index.php" 200
[2021-03-17 13:24:59] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:24:59 +0000 "GET /index.php" 200
[2021-03-17 13:25:00] local.NOTICE:  --- Already busy with refresh
127.0.0.1 -  17/Mar/2021:13:25:00 +0000 "GET /index.php" 200
[2021-03-17 13:25:02] local.NOTICE:  --- Already busy with refresh
`

I don't actually get how the pso-explorer should be working. Does it connect to pso? How does it so? Or should it connect directly to the flasharray(s)?

Best,
Dirk

**Comment by dnix101** [comment]
Hi Dirk,
Thanks for reporting this issue. I believe this is an issue with a recent update, that seems to cause issues in the older versions of Kubernetes. I have made a quick update to the code, would you be willing to check this out? 
You can test the new build by changing the version of the container image for the deployment called `pso-explorer`:
```
containers:
  name: pso-explorer
  image: quay.io/purestorage/pso-explorer:v0.5.9
```
to 
```
containers:
  name: pso-explorer
  image: "quay.io/purestorage/pso-explorer:v0.5.10"
```

**Comment by dnix101** [comment]
With regards to the second part of your question, PSO will connect with Kubernetes to get all the Kubernetes objects, including pods, pvcs, etc. It also retrieves the PSO details. It then uses that information to connect to the management interfaces of the arrays to fetch the capacity and performance information.
So it connects to both the Kubernetes API and the FlashArray/FlashBlade management interface.

**Comment by zapalotta** [comment]
Hi dnix101. Thanks for the quick reply and the explanation and especially the update. It works with the 0.5.10 container.

**Comment by dnix101** [comment]
Perfect! I've just pushed the new release, so the fix in release 0.5.10 is now included in any new install.
